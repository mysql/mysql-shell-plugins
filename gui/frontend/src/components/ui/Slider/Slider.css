 /*
 * Copyright (c) 2020, 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation. The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

 .msg.slider {
     --current-value: 50%;
     --handle-size: 20px;

     display: flex;
     flex-direction: row;

     position: relative;
     min-width: 50px;
     min-height: 10px;
     height: 30px;
     margin: 0px calc(var(--handle-size) / 2);

     user-select: none;
 }

 .msg.slider > .body {
     align-self: center;

     width: 100%;
     height: 10px;

     margin: calc(var(--handle-size) / 2);

     border-radius: 10px;
     border: 1px solid var(--button-border);
     background-color: var(--button-background);

     cursor: pointer;
     overflow: visible;
     user-select: none;
 }

 .msg.slider.vertical {
     flex-direction: column;

     min-width: 10px;
     min-height: 50px;
     width: 30px;

     margin: calc(var(--handle-size) / 2) 0px;
 }

 .msg.slider.vertical > .body {
     width: 10px;
     height: 100%;
 }

 .msg.slider > .body .handle {
     position: absolute;
     left: calc(var(--current-value) - var(--handle-size) / 2);

     width: var(--handle-size);
     height: var(--handle-size);

     border: 2px solid var(--foreground);
     border-radius: 50%;
     background-color: var(--foreground);

     box-shadow: inset 0px 0px 5px rgba(0, 0, 0, 0.5);

     align-self: center;
 }

 .msg.slider.vertical > .body .handle {
     left: auto;
     top: calc(var(--current-value) - var(--handle-size) / 2);
 }

 .msg.slider > .dataInput,
 .msg.slider > .title {
     position: absolute;
     width: max-content;
     min-width: 10px;
 }
