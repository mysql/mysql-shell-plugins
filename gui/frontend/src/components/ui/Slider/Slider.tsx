/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import "./Slider.css";

import React from "react";

import { Component, IComponentProperties, Container, Orientation, PointerEventType } from "..";
import { clampValue } from "../../../utilities/helpers";

export interface ISliderProperties extends IComponentProperties {
    value: number;
    vertical?: boolean;
    handleSize?: number;

    onChange?: (value: number) => void;
}

export class Slider extends Component<ISliderProperties> {

    public static defaultProps = {
        disabled: false,
        vertical: false,
        handleSize: 20,
    };

    private sliderRef = React.createRef<HTMLInputElement>();

    public constructor(props: ISliderProperties) {
        super(props);

        this.addHandledProperties("value", "vertical", "handleSize");

        this.connectEvents("onPointerDown", "onPointerUp");
    }

    public set value(newValue: number) {
        const { onChange } = this.mergedProps;

        newValue = clampValue(newValue, 0, 1);
        this.setState({ currentValue: newValue });

        if (this.sliderRef?.current) {
            this.sliderRef.current.style.setProperty("--current-value", `${100 * newValue}%`);
        }

        onChange?.(newValue);
    }

    public componentDidUpdate(): void {
        const { value, handleSize } = this.mergedProps;

        if (this.sliderRef.current) {
            this.sliderRef.current.style.setProperty("--current-value", `${100 * value}%`);
            this.sliderRef.current.style.setProperty("--handle-size", `${handleSize ?? 0}px`);
        }
    }

    public render(): React.ReactNode {
        const { vertical } = this.mergedProps;
        const className = this.getEffectiveClassNames([
            "slider",
            this.classFromProperty(vertical, "vertical"),
        ]);

        return (
            <div
                ref={this.sliderRef}
                className={className}
                {...this.unhandledProperties}
            >
                <Container // Slider body.
                    className="body"
                    orientation={vertical ? Orientation.TopDown : Orientation.LeftToRight}
                >
                    <div
                        className="handle"
                    />
                </Container>
            </div>
        );
    }

    protected handlePointerEvent(type: PointerEventType, e: React.PointerEvent): boolean {
        switch (type) {
            case PointerEventType.Down: {
                this.updateHandlePosition(e);

                const target = e.currentTarget as HTMLElement;
                target.onpointermove = this.handleItemPointerMove;
                target.setPointerCapture(e.pointerId);

                break;
            }

            case PointerEventType.Up: {
                const target = e.currentTarget as HTMLElement;
                target.onpointermove = null;
                target.releasePointerCapture(e.pointerId);

                break;
            }

            default: {
                break;
            }
        }

        return true;
    }

    private handleItemPointerMove = (e: PointerEvent): void => {
        if (this.sliderRef.current) {
            const { vertical, onChange } = this.mergedProps;

            const bounds = this.sliderRef.current.getBoundingClientRect();
            let value;
            if (vertical) {
                value = clampValue((e.clientY - bounds.y) / bounds.height, 0, 1);
            } else {
                value = clampValue((e.clientX - bounds.x) / bounds.width, 0, 1);
            }

            this.sliderRef.current.style.setProperty("--current-value", `${100 * value}%`);

            onChange?.(value);
        }
    };

    private updateHandlePosition(e: React.PointerEvent): void {
        this.handleItemPointerMove(e.nativeEvent);
    }

}
