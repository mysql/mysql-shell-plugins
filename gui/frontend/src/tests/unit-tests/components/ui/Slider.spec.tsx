/*
 * Copyright (c) 2022, 2025, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { render } from "@testing-library/preact";
import { describe, expect, it, vi } from "vitest";

import { Slider } from "../../../../components/ui/Slider/Slider.js";
import { createRef } from "preact";

describe("Slider component tests", (): void => {

    it("Standard Rendering", () => {
        const { container, unmount } = render(
            <Slider
                id="slider1"
                value={0.3}
            />,
        );
        expect(container).toMatchSnapshot();

        unmount();
    });

    it("Value Change", () => {
        const onChange = vi.fn();

        const sliderRef = createRef<Slider>();
        const { container, unmount } = render(
            <Slider
                ref={sliderRef}
                id="slider1"
                value={0.3}
                onChange={onChange}
            />,
        );

        expect(container).toMatchSnapshot();
        expect(sliderRef.current).toBeDefined();

        sliderRef.current!.value = 0.8;
        expect(onChange).toBeCalledWith(0.8);

        unmount();
    });

});
