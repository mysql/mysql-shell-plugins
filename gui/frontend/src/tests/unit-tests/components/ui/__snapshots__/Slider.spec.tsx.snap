// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider component tests Standard Rendering 1`] = `
<Slider
  disabled={false}
  handleSize={20}
  id="slider1"
  value={0.3}
  vertical={false}
>
  <div
    className="msg slider"
    disabled={false}
    id="slider1"
    onPointerDown={[Function]}
    onPointerUp={[Function]}
  >
    <Container
      className="body"
      orientation="row"
    >
      <div
        className="msg container fixedScrollbar body"
        style={
          Object {
            "alignItems": undefined,
            "flexDirection": "row",
            "flexWrap": undefined,
            "justifyContent": undefined,
          }
        }
      >
        <div
          className="handle"
        />
      </div>
    </Container>
  </div>
</Slider>
`;

exports[`Slider component tests Value Change 1`] = `
<Slider
  disabled={false}
  handleSize={20}
  id="slider1"
  onChange={[MockFunction]}
  value={0.3}
  vertical={false}
>
  <div
    className="msg slider"
    disabled={false}
    id="slider1"
    onPointerDown={[Function]}
    onPointerUp={[Function]}
  >
    <Container
      className="body"
      orientation="row"
    >
      <div
        className="msg container fixedScrollbar body"
        style={
          Object {
            "alignItems": undefined,
            "flexDirection": "row",
            "flexWrap": undefined,
            "justifyContent": undefined,
          }
        }
      >
        <div
          className="handle"
        />
      </div>
    </Container>
  </div>
</Slider>
`;
